Add nodejs binaries pkg blog post

danielrbradley opened the pull request.

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,131 @@
+
+It's well known that you can write Pulumi programs in loads of languages, but what about the providers behind the scenes which handle communication with various cloud services?
+
+Providers are downloaded and run in many different platforms (e.g. Linux, macOS, Windows) and environments – where we can’t assume an end-user is going to have specific runtimes installed – including Node.js. Most of our providers are authored in Go – which makes it simple to produce native, static binaries for all modern operating systems. However, in this post we’re going to look at how we can  ...
```
I'd add a link to vercel/pkg. I wasn't previously familiar with that project.

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
````diff
@@ -0,0 +1,131 @@
+Pkg is distributed as an npm package which can be installed into your “devDependencies” using:
+
+```bash
+npm install -D  pkg
````
```suggestion
npm install -D pkg
```

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,131 @@
+* The target machine to build for
+* The output path to write the finished binary
+
+Here’s an example of building the project in the current directory using node v16 for macOS ARM architectures:
```
Is `v16` correct? The command looks like `v18`

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,131 @@
+
+When trying to use the binaries produced by our CI, we found that [the binaries weren’t runnable on MacOS ARM architectures](https://github.com/pulumi/pulumi-awsx/issues/850) - and were forcibly killed by the operating system.
+
+This led me down quite a bit of a rabbit hole investigating signing of binaries, but it was actually resolved by simply installing the ‘ldid’ tool on our linux CI environments. Sometimes reading the warnings carefully can save you lots of time! The ldid source is available via git://git.saurik.com/ldid.git and binaries are available from various sources. Our solution was to use the [“Install Ldid ...
```
```suggestion
This led me down quite a bit of a rabbit hole investigating signing of binaries, but it was actually resolved by simply installing the ‘ldid’ tool on our linux CI environments. Sometimes reading the warnings carefully can save you lots of time! The ldid source is available via `git://git.saurik.com/ldid.git` and binaries are available from various sources. Our solution was to use the [“Install Ldid” GitHub Action](https://github.com/marketplace/actions/install-ldid) to install the ldid binary and add it to the PATH in our CI workflow.
```

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,131 @@
+meta_image: "meta.png"
+---
+
+It's well known that you can write Pulumi programs in loads of languages, but what about the providers behind the scenes which handle communication with various cloud services?
```
I'd suggest fleshing out the intro section more, and setting some context. This is an engineering deep dive that some readers might opt to skip, so it would be good to cover the high-level points in the first couple of paragraphs, and then transition into the technical details after that.

The questions in my mind right now are:
* What is a provider?
* What is a component provider?
* Why might you want to write a provider in Typescript?

We probably have existing content that covers most of this in detail, so a few links could go a long way here.

danielrbradley replied:
I've made it a little less about providers, because it's not really about them specifically. Hopefully the reworked intro covers more of of the "Why Typescript" and "Why binary packages" part.

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,131 @@
+
+Historical note: we used to do a similar process using the [nexe](https://github.com/nexe/nexe) project, but there’s been no releases since 2017 and therefore no support for newer versions of Node.js and we therefore consider this package as unmaintained at this point in time.
+
+## Setting up pkg
```
Need more of a transition here. What is this section teaching us?

lblackstone reviewed (commented):
You've got a lot of good content here, but I'd suggest working on the intro/conclusion and transitions between sections to tie it all together.

danielrbradley commented:
> You've got a lot of good content here, but I'd suggest working on the intro/conclusion and transitions between sections to tie it all together.

@lblackstone Thanks for the in-depth review - really appreciated! 🙇 

I've rewritten the intro and added a bit more of an outro. Let me know what you think!

danielrbradley commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,131 @@
+3. A wildcard rule invokes pkg build using the TARGET defined above
+4. `$@` is the current makefile target - the specific binary being built
+5. The wildcard rule depends on the node_modules which is also a makefile target - to ensure packages have been restored before the build starts
+6. Define a phony `bins` target to build all bin targets
```
Added a brief recap and sign-off 👍

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,151 @@
+---
+title: "Node.js Native Binary Compilation Using vercel/pkg"
```
I noticed that you use both `vercel/pkg` and `Vercel/pkg` in this post. Which is correct?

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,151 @@
+---
+title: "Node.js Native Binary Compilation Using vercel/pkg"
+date: 2022-09-13
+meta_desc: "Using Vercel pkg to create native binaries for nodejs providers"
```
Oh, here's another difference. We should pick one and use the same terminology consistently.

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,151 @@
+meta_image: "meta.png"
+---
+
+In Pulumi's engineering department we build and distribute our tools as native binaries as there's no local dependencies needed. Go provides a great story for building self-contained binaries that target modern operating systems.
```
```suggestion
In Pulumi's engineering department, we often build and distribute tools as native binaries to avoid the need for additional dependencies on user machines. Most of these tools are written in Go, which has good support for building self-contained binaries that target modern operating systems.
```

lblackstone commented on themes/default/content/blog/nodejs-binaries-with-pkg/index.md:
```diff
@@ -0,0 +1,151 @@
+
+In Pulumi's engineering department we build and distribute our tools as native binaries as there's no local dependencies needed. Go provides a great story for building self-contained binaries that target modern operating systems.
+
+However, it's possible to also author component providers using other languages & runtimes including: Node.js, Python and .NET. However, when it comes to distributing your provider plugin, we'd still like to be able to avoid the user needing to install the required runtime in order to use the provider plugin. In this article we're looking at how we can package a Node.js program into a standalone  ...
```
```suggestion
While other Pulumi-supported languages like Node.js, Python, and .NET require additional runtime dependencies, it's possible to bundle dependencies with the program. In this article, we'll show you how to do that for a Node.js program.
```
